activerecord/lib/active_record/attributes.rb

```diff
@@ -244,7 +244,10 @@ def _default_attributes # :nodoc:
           # TODO: Remove the need for a connection after we release 8.1.
           attributes_hash = with_connection do |connection|
             columns_hash.transform_values do |column|
-              ActiveModel::Attribute.from_database(column.name, column.default, type_for_column(connection, column))
+              type = type_for_column(connection, column)
+              default = column.default
+              default = type.deserialize(default) unless default.nil?
+              ActiveModel::Attribute.from_database(column.name, default, type)
```

Comment:
Would it make sense to deserialize the column default when we build the column instead, so that `Post.columns#default` also returns the deserialized default ?